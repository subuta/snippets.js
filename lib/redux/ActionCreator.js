import { build, format, snippets as s } from 'bld.js'
import _ from 'lodash'
import pluralize from 'pluralize'

import createAction from './Action'
import ActionType from 'lib/redux/ActionType'

const set = ({ tableName }) => {
  // ensure naming convention.
  const models = _.camelCase(pluralize(tableName))

  const ACTION_TYPE = ActionType.set(models)
  const name = _.camelCase(ACTION_TYPE)
  const Action = createAction(ACTION_TYPE, tableName)

  return build`
    const ${name} = (${tableName}) => {
      return ${Action}
    }
  `
}

const create = ({ tableName }) => {
  // ensure naming convention.
  const models = _.camelCase(pluralize(tableName))

  const ACTION_TYPE = ActionType.create(models)
  const name = _.camelCase(ACTION_TYPE)

  const SET_ACTION_TYPE = ActionType.set(models)
  const setAction = _.camelCase(SET_ACTION_TYPE)

  return build`
    const ${name} = (params) => {
      return (dispatch) => {
        dispatch({type: ${ActionType.request(models)}})
        return api.${models}.create(params).then((data) => {
          dispatch(${setAction}(data))
          return data
        })        
      }
    }
  `
}

const show = ({ tableName }) => {
  // ensure naming convention.
  const models = _.camelCase(pluralize(tableName))

  const ACTION_TYPE = ActionType.request(models)
  const name = _.camelCase(ACTION_TYPE)

  const SET_ACTION_TYPE = ActionType.set(models)
  const setAction = _.camelCase(SET_ACTION_TYPE)

  return build`
    const ${pluralize.singular(name)} = (id) => {
      return (dispatch) => {
        dispatch({type: ${ActionType.request(models)}})
        return api.${models}.show(id).then((data) => {
          dispatch(${setAction}(data))
          return data
        })        
      }
    }
  `
}

const all = ({ tableName }) => {
  // ensure naming convention.
  const models = _.camelCase(pluralize(tableName))

  const ACTION_TYPE = ActionType.request(models)
  const name = _.camelCase(ACTION_TYPE)

  const SET_ACTION_TYPE = ActionType.set(models)
  const setAction = _.camelCase(SET_ACTION_TYPE)

  return build`
    const ${name} = () => {
      return (dispatch) => {
        dispatch({type: ${ActionType.request(models)}})
        return api.${models}.index().then((data) => {
          dispatch(${setAction}(data))
          return data
        })        
      }
    }
  `
}

const update = ({ tableName }) => {
  // ensure naming convention.
  const models = _.camelCase(pluralize(tableName))

  const ACTION_TYPE = ActionType.update(models)
  const name = _.camelCase(ACTION_TYPE)

  const SET_ACTION_TYPE = ActionType.set(models)
  const setAction = _.camelCase(SET_ACTION_TYPE)

  return build`
    const ${pluralize.singular(name)} = (params, id) => {
      return (dispatch) => {
        dispatch({type: ${ActionType.request(models)}})
        return api.${models}.update(params, id).then((data) => {
          dispatch(${setAction}(data))
          return data
        })        
      }
    }
  `
}

const destroy = ({ tableName }) => {
  // ensure naming convention.
  const models = _.camelCase(pluralize(tableName))

  const ACTION_TYPE = ActionType.destroy(models)
  const name = _.camelCase(ACTION_TYPE)

  const SET_ACTION_TYPE = ActionType.set(models)
  const setAction = _.camelCase(SET_ACTION_TYPE)

  return build`
    const ${pluralize.singular(name)} = (id) => {
      return (dispatch) => {
        dispatch({type: ${ActionType.request(models)}})
        return api.${models}.destroy(id).then(() => {
          dispatch(${setAction}({}))
        })        
      }
    }
  `
}

export default {
  set,
  create,
  show,
  all,
  update,
  destroy
}