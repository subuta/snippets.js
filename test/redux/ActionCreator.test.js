import test from 'ava'
import { build, format, snippets as s } from 'bld.js'

import ActionCreator from 'lib/redux/ActionCreator'

test.beforeEach(async (t) => {

})

test.afterEach((t) => {
})

test('should create redux ActionCreator for setter', async (t) => {
  const code = build`
    ${ActionCreator.set({ tableName: 'books' })}
  `

  const expected = build`
    const setBooks = (books) => {
      return {
        type: SET_BOOKS,
        payload: books
      }
    }

  `

  t.is(format(code), format(expected))
})