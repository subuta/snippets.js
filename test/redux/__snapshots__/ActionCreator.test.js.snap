// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux/ActionCreator should create redux ActionCreator for all(index) Action 1`] = `
"const requestBooks = () => {
  return (dispatch) => {
    dispatch({type: REQUEST_BOOKS})
    return api.book.index().then((data) => {
      /* mat Index data transform [start] */
      /* mat Index data transform [end] */
      dispatch(setBooks(normalize(data, bookList)))
      return data
    })
  }
}
"
`;

exports[`redux/ActionCreator should create redux ActionCreator for create Action 1`] = `
"const createBook = (params) => {
  return (dispatch) => {
    dispatch({type: REQUEST_BOOKS})
    return api.book.create(params).then((data) => {
      /* mat Create data transform [start] */
      /* mat Create data transform [end] */
      dispatch(setBooks(normalize(data, book)))
      return data
    })
  }
}
"
`;

exports[`redux/ActionCreator should create redux ActionCreator for destroy Action 1`] = `
"const deleteBook = (id) => {
  return (dispatch) => {
    dispatch({type: REQUEST_BOOKS})
    return api.book.destroy(id).then(() => {
      dispatch(setBooks({}))
    })
  }
}
"
`;

exports[`redux/ActionCreator should create redux ActionCreator for setter 1`] = `
"const setBooks = (books) => {
  return {
    type: SET_BOOKS,
    payload: books
  }
}
"
`;

exports[`redux/ActionCreator should create redux ActionCreator for show Action 1`] = `
"const requestBook = (id) => {
  return (dispatch) => {
    dispatch({type: REQUEST_BOOKS})
    return api.book.show(id).then((data) => {
      /* mat Show data transform [start] */
      /* mat Show data transform [end] */
      dispatch(setBooks(normalize(data, book)))
      return data
    })
  }
}
"
`;

exports[`redux/ActionCreator should create redux ActionCreator for update Action 1`] = `
"const updateBook = (id, params) => {
  return (dispatch) => {
    dispatch({type: REQUEST_BOOKS})
    return api.book.update(id, params).then((data) => {
      /* mat Update data transform [start] */
      /* mat Update data transform [end] */
      dispatch(setBooks(normalize(data, book)))
      return data
    })
  }
}
"
`;
